extensions/azurecore/src/account-provider/auths/azureAuth.ts

```diff
@@ -0,0 +1,491 @@
+	/**
+	 * Access Token
+	 */
+	token: string;
+}
+
+export interface RefreshToken extends AccountKey {
+	/**
+	 * Refresh Token
+	 */
+	token: string;
+
+	/**
+	 * Account Key
+	 */
+	key: string
+}
+
+export interface TokenResponse {
+	[tenantId: string]: Token
+}
+
+export interface Token extends AccountKey {
+	/**
+	 * Access token
+	 */
+	token: string;
+
+	/**
+	 * TokenType
+	 */
+	tokenType: string;
+}
+
+export interface TokenClaims { // https://docs.microsoft.com/en-us/azure/active-directory/develop/id-tokens
+	aud: string;
+	iss: string;
+	iat: number;
+	idp: string,
+	nbf: number;
+	exp: number;
+	c_hash: string;
+	at_hash: string;
+	aio: string;
+	preferred_username: string;
+	email: string;
+	name: string;
+	nonce: string;
+	oid: string;
+	roles: string[];
+	rh: string;
+	sub: string;
+	tid: string;
+	unique_name: string;
+	uti: string;
+	ver: string;
+}
+
+export type TokenRefreshResponse = { accessToken: AccessToken, refreshToken: RefreshToken, tokenClaims: TokenClaims };
+
+export abstract class AzureAuth {
+	protected readonly memdb = new MemoryDatabase();
+
+	protected readonly WorkSchoolAccountType: string = 'work_school';
+	protected readonly MicrosoftAccountType: string = 'microsoft';
+
+	protected readonly loginEndpointUrl: string;
+	protected readonly commonTenant: string;
+	protected readonly redirectUri: string;
+	protected readonly scopes: string[];
+	protected readonly scopesString: string;
+	protected readonly clientId: string;
+	protected readonly resources: Resource[];
+
+
+	constructor(
+		protected readonly metadata: AzureAccountProviderMetadata,
+		protected readonly tokenCache: SimpleTokenCache,
+		protected readonly context: vscode.ExtensionContext,
+		protected readonly authType: AzureAuthType,
+		public readonly userFriendlyName: string
+	) {
+		this.loginEndpointUrl = this.metadata.settings.host;
+		this.commonTenant = 'common';
+		this.redirectUri = this.metadata.settings.redirectUri;
+		this.clientId = this.metadata.settings.clientId;
+
+		this.resources = [
+			this.metadata.settings.armResource, this.metadata.settings.sqlResource,
+			this.metadata.settings.graphResource, this.metadata.settings.ossRdbmsResource
+		];
+
+		this.scopes = [...this.metadata.settings.scopes];
+		this.scopesString = this.scopes.join(' ');
+	}
+
+	public abstract async login(): Promise<AzureAccount | azdata.PromptFailedResult>;
+
+	public abstract async autoOAuthCancelled(): Promise<void>;
+
+
+	public async refreshAccess(account: azdata.Account): Promise<azdata.Account> {
+		const response = await this.getCachedToken(account.key);
+		if (!response) {
+			account.isStale = true;
+			return account;
+		}
+
+		const refreshToken = response.refreshToken;
+		if (!refreshToken || !refreshToken.key) {
+			account.isStale = true;
+			return account;
+		}
+
+		try {
+			await this.refreshAccessToken(account.key, refreshToken);
+		} catch (ex) {
+			if (ex.message) {
+				await vscode.window.showErrorMessage(ex.message);
```

Comment:
Man, I really think we need like a "LocalizedError" type or something. Most of the time we don't want error messages localized since when they show up in logs we want them in English. But then we want to be able to display them in situations like this too...

Reply:
That would be perfect...